modules/packages/Channel.chpl

```diff
@@ -0,0 +1,559 @@
+                    recvidx = (recvidx + 1) % bufferSize;
+                }
+                else val = sender.val.deref();
+
+                sender.release(true);
+            }
+            else {
+
+                recvidx = (recvidx + 1) % bufferSize;
+                count -= 1;
+
+            }
+            if !selected then unlock();
+
+            return true;
+
+        }
+
+        /*
+        Send a value to the channel buffer. If the buffer is at maximum
+        capacity it will suspend the waiting task, until there is space in the
+        buffer or a receiving task awakes it. If a channel is closed no more
+        data can be sent to it.
+
+        :arg val: Data to be sent to the channel
+        :type val: `eltType`
+
+        :throws ChannelError: If ``send`` is called on a closed channel.
+        */
+
+        proc send(in val : eltType) throws {
+            send(val, false);
+        }
+
+        pragma "no doc"
+        proc send(in val : eltType, selected : bool) : bool throws {
+            if !selected then lock();
+
+            if closed {
+                if !selected then unlock();
+                throw new owned ChannelError("Sending on a closed channel");
+            }
+
+            while !recvWaiters.isEmpty() && recvWaiters.front!.isSelect {
+                if recvWaiters.front!.isSelectDone.deref().compareAndSwap(-1, recvWaiters.front!.selID) {
+                    break;
+                }
+                else recvWaiters.deque();
+            }
+
+            if count == bufferSize && recvWaiters.isEmpty() {
+                if selected then return false;
+                var process$ : single bool;
+                var processing = new unmanaged Waiter(val, process$);
+
+                sendWaiters.enque(processing);
+                unlock();
+                var status = processing.suspend();
+                delete processing;
+                if status == false {
+                    throw new owned ChannelError("Sending on a closed channel");
+                }
+                return status;
+
+            }
+            else {
+                if !recvWaiters.isEmpty() {
+                    var receiver = recvWaiters.deque();
+                    receiver.val.deref() = val;
+
+                    receiver.release(true);
+                }
+                else {
+                    buffer[sendidx] = val;
+
+                    sendidx = (sendidx + 1) % bufferSize;
+                    count += 1;
+                }
+
+                if !selected then unlock();
+                return true;
+            }
+        }
+
+        /*
+        This function is used to close a channel indicating that no more data
+        can be sent to it.
+
+        :throws ChannelError: If called on a closed channel.
+        */
+
+        proc close() throws {
+
+            lock();
+            if closed {
+                unlock();
+                throw new owned ChannelError("Closing a closed channel");
```

Comment:
Can we rephrase this error message to something better -> like `Trying to close a closed channel` or something even more better you can think of?